homeassistant/components/minio/__init__.py

```diff
@@ -0,0 +1,280 @@
+CONF_LISTEN = 'listen'
+CONF_LISTEN_BUCKET = 'bucket'
+CONF_LISTEN_PREFIX = 'prefix'
+CONF_LISTEN_SUFFIX = 'suffix'
+CONF_LISTEN_EVENTS = 'events'
+
+CONF_LISTEN_PREFIX_DEFAULT = ''
+CONF_LISTEN_SUFFIX_DEFAULT = '.*'
+CONF_LISTEN_EVENTS_DEFAULT = 's3:ObjectCreated:*'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PORT): cv.port,
+        vol.Required(CONF_ACCESS_KEY): cv.string,
+        vol.Required(CONF_SECRET_KEY): cv.string,
+        vol.Required(CONF_SECURE): cv.boolean,
+        vol.Optional(CONF_LISTEN): vol.All(cv.ensure_list, [vol.Schema({
+            vol.Required(CONF_LISTEN_BUCKET): cv.string,
+            vol.Optional(
+                CONF_LISTEN_PREFIX,
+                default=CONF_LISTEN_PREFIX_DEFAULT
+            ): cv.string,
+            vol.Optional(
+                CONF_LISTEN_SUFFIX,
+                default=CONF_LISTEN_SUFFIX_DEFAULT
+            ): cv.string,
+            vol.Optional(
+                CONF_LISTEN_EVENTS,
+                default=CONF_LISTEN_EVENTS_DEFAULT
+            ): cv.string,
+        })])
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+BUCKET_KEY_SCHEMA = vol.Schema({
+    vol.Required('bucket'): cv.template,
+    vol.Required('key'): cv.template,
+})
+
+BUCKET_KEY_FILE_SCHEMA = BUCKET_KEY_SCHEMA.extend({
+    vol.Required('file_path'): cv.template,
+})
+
+
+def get_minio_endpoint(host: str, port: int) -> str:
+    """Create minio endpoint from host and port."""
+    return host + ':' + str(port)
+
+
+class QueueListener(threading.Thread):
+    """Forward events from queue into HASS event bus."""
+
+    def __init__(self, hass):
+        """Create queue."""
+        super().__init__()
+        self.__hass = hass
+        self.__q = Queue()
+
+    def run(self):
+        """Listen to queue events, and forward them to HASS event bus."""
+        _LOGGER.info('Running QueueListener')
+        while True:
+            event = self.__q.get()
+            if event is None:
+                break
+
+            _, file_name = os.path.split(event['key'])
+
+            _LOGGER.debug(
+                'Sending event %s, %s, %s',
+                event['event_name'],
+                event['bucket'],
+                event['key']
+            )
+            self.__hass.bus.fire(DOMAIN, {
+                'file_name': file_name,
+                **event,
+            })
+
+    @property
+    def queue(self):
+        """Return wrapped queue."""
+        return self.__q
+
+    def stop(self):
+        """Stop run by putting None into queue and join the thread."""
+        _LOGGER.info('Stopping QueueListener')
+        self.__q.put(None)
+        self.join()
+        _LOGGER.info('Stopped QueueListener')
+
+    def start_handler(self, _):
+        """Start handler helper method."""
+        self.start()
+
+    def stop_handler(self, _):
+        """Stop handler helper method."""
+        self.stop()
+
+
+class MinioListener:
```

Comment:
Please move this below `setup` function. We want the `setup` function close to the config schema for easy reading.